cpp/test/sg/fil_test.cu

```diff
@@ -196,29 +249,79 @@ class BaseFilTest : public testing::TestWithParam<FilTestParams> {
     r.uniformInt(fids_d, num_nodes, 0, ps.num_cols, stream);
     r.bernoulli(def_lefts_d, num_nodes, 0.5f, stream);
     r.bernoulli(is_leafs_d, num_nodes, 1.0f - ps.leaf_prob, stream);
+    hard_clipped_bernoulli(
+      r, is_categoricals_d, num_nodes, 1.0f - ps.node_categorical_prob, stream);
 
     // copy data to host
-    std::vector<float> thresholds_h(num_nodes);
-    std::vector<int> weights_h(num_nodes), fids_h(num_nodes);
+    std::vector<float> thresholds_h(num_nodes), is_categoricals_h(num_nodes);
+    std::vector<int> max_matching_cat_h(ps.num_cols), weights_h(num_nodes), fids_h(num_nodes),
+      node_cat_set(num_nodes);
+    std::vector<bool> feature_categorical(ps.num_cols);
+    // bool vectors are not guaranteed to be stored byte-per-value
     def_lefts_h = new bool[num_nodes];
     is_leafs_h  = new bool[num_nodes];
 
+    // uniformily distributed in orders of magnitude: smaller models which
+    // still stress large bitfields.
+    // up to 10**ps.max_matching_cat_oom (only if feature is categorical, else -1)
+    std::vector<cat_feature_counters> cf(ps.num_cols);
+    std::mt19937 gen(ps.seed);
+    std::uniform_real_distribution mmc(-1.0f, ps.max_matching_cat_oom);
```

Comment:
What about `mmc(0.0f, ps.max_matching_cat_oom + 1.0f)`? Then you don't need to add `1.0f`.